plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php

```diff
@@ -0,0 +1,186 @@
+					$where_ids[] = absint( $arg );
+				} elseif ( preg_match( '/^(\d+)-(\d+)$/', $arg, $matches ) ) {
+					$where_ranges[] = $wpdb->prepare( "({$wpdb->posts}.ID >= %d AND {$wpdb->posts}.ID <= %d)", absint( $matches[1] ), absint( $matches[2] ) );
+				}
+			}
+
+			if ( $where_ids ) {
+				$where_ranges[] = "{$wpdb->posts}.ID IN (" . implode( ',', $where_ids ) . ')';
+			}
+
+			if ( ! $where_ranges ) {
+				$sql_where .= '1=0';
+			} else {
+				$sql_where .= '(' . implode( ' OR ', $where_ranges ) . ')';
+			}
+		}
+
+		$sql_where .= $sql_where ? ' AND ' : '';
+		$sql_where .= "{$wpdb->posts}.post_type IN ('" . implode( "', '", esc_sql( wc_get_order_types( 'cot-migration' ) ) ) . "')";
+
+		if ( 'count' === $result ) {
+			$sql_fields = 'COUNT(*)';
+			$sql_limit  = '';
+		} else {
+			$sql_fields = 'ID';
+			$sql_limit  = $limit > 0 ? $wpdb->prepare( 'LIMIT %d', $limit ) : '';
+		}
+
+		return "SELECT {$sql_fields} FROM {$wpdb->posts} WHERE {$sql_where} {$sql_limit}";
+	}
+
+	/**
+	 * Performs a cleanup of post data for a given order and also converts the post to the placeholder type in the backup table.
+	 *
+	 * @param int $order_id Order ID.
+	 * @return void
+	 * @throws \Exception When an error occurs.
+	 */
+	public function cleanup_post_data( $order_id ): void {
+		global $wpdb;
+
+		$order = wc_get_order( $order_id );
+		if ( ! $order ) {
+			// translators: %d is an order ID.
+			throw new \Exception( sprintf( __( '%d is not a valid order ID.', 'woocommerce' ), $order_id ) );
+		}
+
+		if ( ! $this->is_order_newer_than_post( $order ) ) {
+			throw new \Exception( sprintf( __( 'Data in posts table appears to be more recent than in HPOS tables.', 'woocommerce' ) ) );
+		}
+
+		$meta_ids = $wpdb->get_col( $wpdb->prepare( "SELECT meta_id FROM {$wpdb->postmeta} WHERE post_id = %d", $order->get_id() ) );
+		foreach ( $meta_ids as $meta_id ) {
+			delete_metadata_by_mid( 'post', $meta_id );
+		}
+
+		$wpdb->query(
+			$wpdb->prepare(
+				"UPDATE {$wpdb->posts} SET post_type = %s, post_status = %s WHERE ID = %d",
+				$this->data_synchronizer::PLACEHOLDER_ORDER_POST_TYPE,
+				'draft',
+				$order->get_id()
+			)
+		);
+
+		clean_post_cache( $order->get_id() );
+	}
+
+	/**
+	 * Checks whether an HPOS-backed order is newer than the corresponding post.
+	 *
+	 * @param int|\WC_Order $order An HPOS order.
+	 * @return bool TRUE if the order is up to date with the corresponding post.
+	 * @throws \Exception When the order is not an HPOS order.
+	 */
+	private function is_order_newer_than_post( $order ): bool {
+		$order = is_a( $order, 'WC_Order' ) ? $order : wc_get_order( absint( $order ) );
+
+		if ( ! is_a( $order->get_data_store()->get_current_class_name(), OrdersTableDataStore::class, true ) ) {
+			throw new \Exception( __( 'Order is not an HPOS order.', 'woocommerce' ) );
+		}
+
+		$post = get_post( $order->get_id() );
+		if ( ! $post || $this->data_synchronizer::PLACEHOLDER_ORDER_POST_TYPE === $post->post_type ) {
```

Comment:
I thought about this a bit more, and IMHO cleaning up placeholder posts that have metadata should be supported. In theory, those posts should always remain free of meta.

As such, I've modified the queries (see 3b92b521cb4a8f1cec3ccea950dc9385fa6db37f), so that both regular order types _and_ placeholders (with meta) are considered by the cleanup code.

I still decided to add support for the `--force` flag (in 0642bcbf5720d0cadf795796470fae9432714605) but it works as a way to skip the checks, which right now are just making sure that the post is not more recent than the order itself, though might be more comprehensive in the future.

What do you think of this approach?

Reply:
:+1: